File: backend/simulator/commands/merge.py
from __future__ import annotations

import copy

from simulator.commands.base import BaseCommandHandler, CommandOutcome, SimulatorCommandError
from simulator.intents import CommandIntent


class MergeCommandHandler(BaseCommandHandler):
    command = "merge"

    def apply(self, runtime, state: dict, intent: CommandIntent) -> CommandOutcome:
        operation = intent.operations[0]
        if operation.name == "AbortMerge":
            return self._abort(runtime, state)
        if operation.name == "ContinueMerge":
            return self._continue(runtime, state)
        return self._merge(runtime, state, operation.params["branch"])

    def _abort(self, runtime, state: dict) -> CommandOutcome:
        if not state.get("merge_parent") and not state.get("conflicts"):
            raise SimulatorCommandError("fatal: There is no merge to abort (MERGE_HEAD missing).")
        restored = state.get("merge_abort_state")
        if isinstance(restored, dict) and restored:
            state.clear()
            state.update(copy.deepcopy(restored))
            runtime._set_operation_metadata(state, last_merge_aborted=True)
            return CommandOutcome(command="merge", stdout="")
        state["conflicts"] = []
        state["staging"] = {}
        state["working_tree"] = {}
        state.pop("merge_parent", None)
        runtime._set_operation_metadata(state, last_merge_aborted=True)
        return CommandOutcome(command="merge", stdout="")

    def _continue(self, runtime, state: dict) -> CommandOutcome:
        if not state.get("merge_parent"):
            raise SimulatorCommandError("fatal: There is no merge in progress (MERGE_HEAD missing).")
        if state.get("conflicts"):
            raise SimulatorCommandError(
                "Committing is not possible because you have unmerged files.",
                exit_code=128,
            )
        from simulator.commands.commit import CommitCommandHandler

        return CommitCommandHandler()._create(runtime, state, {"message": None}, staged_by_all=[])

    def _merge(self, runtime, state: dict, branch: str) -> CommandOutcome:
        if state.get("merge_parent") or state.get("conflicts"):
            raise SimulatorCommandError(
                "error: Merging is not possible because you have unmerged files.",
                exit_code=128,
            )
        target_id = self._resolve_ref(state, branch)
        if not target_id:
            raise SimulatorCommandError(f"merge: {branch} - not something we can merge")
        current_id = runtime._head_commit(state)
        if current_id == target_id:
            return CommandOutcome(command="merge", stdout="Already up to date.")

        before_merge = copy.deepcopy(state)
        current_tree = runtime._tree_for_commit(state, current_id)
        target_tree = runtime._tree_for_commit(state, target_id)
        base_id = self._common_ancestor(state, current_id, target_id)
        base_tree = runtime._tree_for_commit(state, base_id)
        conflict_paths = set(self._authored_conflict_paths(state))
        conflict_paths.update(self._overlapping_conflicts(base_tree, current_tree, target_tree))

        staged_paths: list[str] = []
        for path in sorted(set(current_tree) | set(target_tree)):
            if path in conflict_paths or current_tree.get(path) == target_tree.get(path):
                continue
            after = target_tree.get(path)
            state.setdefault("staging", {})[path] = (
                "deleted"
                if after is None
                else {
                    "status": runtime.normalizer.change_type(current_tree.get(path), after),
                    "content": copy.deepcopy(after),
                }
            )
            staged_paths.append(path)

        if conflict_paths:
            state["merge_parent"] = target_id
            state["merge_abort_state"] = before_merge
            state["conflicts"] = sorted(conflict_paths)
            for path in sorted(conflict_paths):
                state.setdefault("working_tree", {})[path] = self._conflict_entry(
                    path=path,
                    branch=branch,
                    base=base_tree.get(path),
                    ours=current_tree.get(path),
                    theirs=target_tree.get(path),
                    resolution=(state.get("merge_resolutions") or {}).get(path),
                )
            runtime._set_operation_metadata(
                state,
                last_merge_branch=branch,
                last_merge_target=target_id,
                last_merge_conflicted=True,
                last_merge_conflict_paths=sorted(conflict_paths),
            )
            return CommandOutcome(
                command="merge",
                details={"branch": branch, "conflicts": sorted(conflict_paths)},
                exit_code=1,
                stdout=(
                    f"Auto-merging {', '.join(sorted(conflict_paths))}\n"
                    "CONFLICT (content): Merge conflict in "
                    f"{', '.join(sorted(conflict_paths))}\n"
                    "Automatic merge failed; fix conflicts and then commit the result."
                ),
            )

        if self._is_ancestor(state, current_id, target_id):
            runtime._set_head_commit(state, target_id)
            runtime._set_operation_metadata(
                state,
                last_merge_branch=branch,
                last_merge_target=target_id,
                last_merge_fast_forward=True,
            )
            return CommandOutcome(command="merge", stdout=f"Fast-forward\nMerged {branch}.")

        commit_id = runtime._next_commit_id(state)
        changes = runtime._diff_trees(current_tree, target_tree)
        state.setdefault("commits", []).append(
            runtime._commit_payload(
                state=state,
                commit_id=commit_id,
                message=f"Merge branch '{branch}'",
                parents=[item for item in [current_id, target_id] if item],
                tree=target_tree,
                changes=changes,
            )
        )
        state["staging"] = {}
        runtime._set_head_commit(state, commit_id)
        runtime._set_operation_metadata(
            state,
            last_merge_branch=branch,
            last_merge_target=target_id,
            last_merge_created_commit=commit_id,
            last_merge_auto_staged_paths=staged_paths,
        )
        return CommandOutcome(command="merge", stdout=f"Merge made by the 'ort' strategy.\n {len(changes)} file(s) changed")

    def _resolve_ref(self, state: dict, ref: str) -> str | None:
        if ref in state.get("branches", {}):
            return state["branches"][ref]
        if ref in state.get("remote_branches", {}):
            return state["remote_branches"][ref]
        return ref if any(commit.get("id") == ref for commit in state.get("commits", [])) else None

    def _authored_conflict_paths(self, state: dict) -> list[str]:
        if not state.get("conflict_on_merge") and not state.get("merge_conflicts"):
            return []
        paths = state.get("conflict_files") or state.get("merge_conflict_files") or []
        if isinstance(state.get("merge_conflicts"), dict):
            paths = [*paths, *state["merge_conflicts"].keys()]
        return sorted(str(path) for path in paths)

    def _overlapping_conflicts(self, base_tree: dict, current_tree: dict, target_tree: dict) -> set[str]:
        conflicts = set()
        for path in set(base_tree) | set(current_tree) | set(target_tree):
            base = base_tree.get(path)
            ours = current_tree.get(path)
            theirs = target_tree.get(path)
            if ours != base and theirs != base and ours != theirs:
                conflicts.add(path)
        return conflicts

    def _conflict_entry(
        self,
        *,
        path: str,
        branch: str,
        base: object,
        ours: object,
        theirs: object,
        resolution: object,
    ) -> dict:
        ours_text = "" if ours is None else str(ours)
        theirs_text = "" if theirs is None else str(theirs)
        return {
            "status": "conflicted",
            "content": f"<<<<<<< HEAD\n{ours_text}\n=======\n{theirs_text}\n>>>>>>> {branch}\n",
            "base": base,
            "ours": ours,
            "theirs": theirs,
            "resolution": resolution,
        }

    def _common_ancestor(self, state: dict, left: str | None, right: str | None) -> str | None:
        left_history = self._history(state, left)
        right_history = set(self._history(state, right))
        return next((commit_id for commit_id in left_history if commit_id in right_history), None)

    def _is_ancestor(self, state: dict, ancestor: str | None, descendant: str | None) -> bool:
        return bool(ancestor and ancestor in self._history(state, descendant))

    def _history(self, state: dict, commit_id: str | None) -> list[str]:
        commits = {commit["id"]: commit for commit in state.get("commits", [])}
        stack = [commit_id] if commit_id else []
        seen: list[str] = []
        while stack:
            current = stack.pop()
            if current in seen or current not in commits:
                continue
            seen.append(current)
            stack.extend(commits[current].get("parents", []))
        return seen

File: frontend/src/shared/git/simulator/commands/merge.ts
import type { RepositoryValue } from '@/shared/level/types'
import { hasOption, optionValues } from '@/shared/git/simulator/parser'
import { applyChanges, changeType, clone, commitById, commitPayload, diffTrees, headBranch, headCommit, isRecord, nextCommitId, setHeadCommit, setOperationMetadata, treeForCommit } from '@/shared/git/simulator/state'
import { createCommit } from '@/shared/git/simulator/commands/commits'
import { commonAncestor, commitsSince, isAncestor, resolveRef } from '@/shared/git/simulator/commands/refs'
import { SimulatorCommandError, type CommandOutcome, type MutableRepositoryState, type ParsedGitCommand } from '@/shared/git/simulator/types'

export function gitMerge(state: MutableRepositoryState, parsed: ParsedGitCommand): CommandOutcome {
  if (hasOption(parsed, '--abort')) return abortMerge(state)
  if (hasOption(parsed, '--continue')) return continueMerge(state)
  if (hasOption(parsed, '--squash')) return squashMerge(state, parsed.args[0])
  return mergeBranch(state, parsed.args[0], hasOption(parsed, '--no-ff'))
}

function abortMerge(state: MutableRepositoryState): CommandOutcome {
  if (!state.merge_parent && !state.conflicts?.length) throw new SimulatorCommandError('fatal: There is no merge to abort (MERGE_HEAD missing).')
  const restored = state.merge_abort_state
  if (restored && isRecord(restored) && Object.keys(restored).length > 0) {
    for (const key of Object.keys(state)) delete state[key]
    Object.assign(state, clone(restored))
    setOperationMetadata(state, { last_merge_aborted: true })
    return { command: 'merge', stdout: '' }
  }
  state.conflicts = []
  state.staging = {}
  state.working_tree = {}
  delete state.conflict_details
  delete state.merge_parent
  setOperationMetadata(state, { last_merge_aborted: true })
  return { command: 'merge', stdout: '' }
}

function continueMerge(state: MutableRepositoryState) {
  if (!state.merge_parent) throw new SimulatorCommandError('fatal: There is no merge in progress (MERGE_HEAD missing).')
  if (state.conflicts?.length) throw new SimulatorCommandError('Committing is not possible because you have unmerged files.')
  return createCommit(state, null, [])
}

function squashMerge(state: MutableRepositoryState, branch: string): CommandOutcome {
  if (state.merge_parent || state.conflicts?.length) throw new SimulatorCommandError('error: Merging is not possible because you have unmerged files.')
  const targetId = resolveRef(state, branch)
  if (!targetId) throw new SimulatorCommandError(`merge: ${branch} - not something we can merge`)
  const currentId = headCommit(state)
  if (currentId === targetId) return { command: 'merge', stdout: 'Already up to date.' }
  const currentTree = treeForCommit(state, currentId)
  const targetTree = treeForCommit(state, targetId)
  state.staging ??= {}
  for (const [path, value] of Object.entries(targetTree)) {
    if (JSON.stringify(value) !== JSON.stringify(currentTree[path])) {
      state.staging[path] = { status: changeType(currentTree[path] ?? null, value), content: clone(value ?? null) }
    }
  }
  for (const path of Object.keys(currentTree)) {
    if (!(path in targetTree)) state.staging[path] = 'deleted'
  }
  setOperationMetadata(state, { last_merge_branch: branch, last_merge_target: targetId, squash_merge_staged: true })
  return { command: 'merge', stdout: 'Squash commit -- not updating HEAD\nAutomatic merge went well; stopped before committing as requested' }
}

export function mergeBranch(state: MutableRepositoryState, branch: string, noFf: boolean): CommandOutcome {
  if (state.merge_parent || state.conflicts?.length) throw new SimulatorCommandError('error: Merging is not possible because you have unmerged files.')
  const targetId = resolveRef(state, branch)
  if (!targetId) throw new SimulatorCommandError(`merge: ${branch} - not something we can merge`)
  const currentId = headCommit(state)
  if (currentId === targetId) return { command: 'merge', stdout: 'Already up to date.' }

  const beforeMerge = clone(state)
  const currentTree = treeForCommit(state, currentId)
  const targetTree = treeForCommit(state, targetId)
  const baseId = commonAncestor(state, currentId, targetId)
  const baseTree = treeForCommit(state, baseId)
  const conflictPaths = new Set(authoredConflictPaths(state))
  for (const path of overlappingConflicts(baseTree, currentTree, targetTree)) conflictPaths.add(path)

  const stagedPaths: string[] = []
  for (const path of [...new Set([...Object.keys(currentTree), ...Object.keys(targetTree)])].sort()) {
    if (conflictPaths.has(path) || JSON.stringify(currentTree[path]) === JSON.stringify(targetTree[path])) continue
    const after = targetTree[path] ?? null
    state.staging ??= {}
    state.staging[path] = after === null ? 'deleted' : { status: changeType(currentTree[path] ?? null, after), content: clone(after) }
    stagedPaths.push(path)
  }

  if (conflictPaths.size) {
    state.merge_parent = targetId
    state.merge_abort_state = beforeMerge
    state.conflicts = [...conflictPaths].sort()
    state.conflict_details ??= {}
    state.working_tree ??= {}
    for (const path of state.conflicts) {
      state.conflict_details[path] = {
        base: clone(baseTree[path] ?? null),
        ours: clone(currentTree[path] ?? null),
        theirs: clone(targetTree[path] ?? null),
        resolution: clone(state.merge_resolutions?.[path] ?? null),
        merge_branch: branch,
      }
      state.working_tree[path] = conflictEntry(branch, baseTree[path] ?? null, currentTree[path] ?? null, targetTree[path] ?? null, state.merge_resolutions?.[path] ?? null)
    }
    setOperationMetadata(state, { last_merge_branch: branch, last_merge_target: targetId, last_merge_conflicted: true, last_merge_conflict_paths: state.conflicts })
    return {
      command: 'merge',
      details: { branch, conflicts: state.conflicts },
      exitCode: 1,
      stdout: `Auto-merging ${state.conflicts.join(', ')}\nCONFLICT (content): Merge conflict in ${state.conflicts.join(', ')}\nAutomatic merge failed; fix conflicts and then commit the result.`,
    }
  }

  if (!noFf && isAncestor(state, currentId, targetId)) {
    // A fast-forward only moves the current branch ref. The path differences
    // calculated above describe the new commit tree; they must not remain as
    // staged changes after the ref moves.
    state.staging = {}
    state.working_tree = {}
    setHeadCommit(state, targetId)
    setOperationMetadata(state, { last_merge_branch: branch, last_merge_target: targetId, last_merge_fast_forward: true })
    return { command: 'merge', stdout: `Fast-forward\nMerged ${branch}.` }
  }

  const commitId = nextCommitId(state)
  const changes = diffTrees(currentTree, targetTree)
  state.commits ??= []
  state.commits.push(commitPayload({ state, commitId, message: `Merge branch '${branch}'`, parents: [currentId, targetId].filter(Boolean) as string[], tree: targetTree, changes }))
  state.staging = {}
  setHeadCommit(state, commitId)
  setOperationMetadata(state, {
    last_merge_branch: branch,
    last_merge_target: targetId,
    last_merge_created_commit: commitId,
    last_merge_auto_staged_paths: stagedPaths,
    last_merge_no_ff: noFf,
  })
  const strategy = noFf ? 'ort (no fast-forward)' : 'ort'
  return { command: 'merge', stdout: `Merge made by the '${strategy}' strategy.\n ${Object.keys(changes).length} file(s) changed` }
}

function conflictEntry(branch: string, _base: RepositoryValue, ours: RepositoryValue, theirs: RepositoryValue, resolution: RepositoryValue) {
  return {
    status: 'conflicted',
    content: `<<<<<<< HEAD\n${ours ?? ''}\n=======\n${theirs ?? ''}\n>>>>>>> ${branch}\n`,
    base: _base,
    ours,
    theirs,
    resolution,
  }
}

function authoredConflictPaths(state: MutableRepositoryState) {
  if (!state.conflict_on_merge && !state.merge_conflicts) return []
  let paths = [...(state.conflict_files ?? []), ...(state.merge_conflict_files ?? [])]
  if (isRecord(state.merge_conflicts)) paths = [...paths, ...Object.keys(state.merge_conflicts)]
  return [...new Set(paths.map(String))].sort()
}

function overlappingConflicts(baseTree: Record<string, RepositoryValue>, currentTree: Record<string, RepositoryValue>, targetTree: Record<string, RepositoryValue>) {
  const conflicts = new Set<string>()
  for (const path of new Set([...Object.keys(baseTree), ...Object.keys(currentTree), ...Object.keys(targetTree)])) {
    const base = JSON.stringify(baseTree[path] ?? null)
    const ours = JSON.stringify(currentTree[path] ?? null)
    const theirs = JSON.stringify(targetTree[path] ?? null)
    if (ours !== base && theirs !== base && ours !== theirs) conflicts.add(path)
  }
  return conflicts
}

export function gitMergetool(state: MutableRepositoryState, parsed: ParsedGitCommand): CommandOutcome {
  const conflicts = state.conflicts ?? []
  if (!conflicts.length) throw new SimulatorCommandError('No files need merging')
  const requested = parsed.pathspecs.length ? parsed.pathspecs : conflicts
  const selected = conflicts.filter((path) => requested.includes(path))
  if (!selected.length) throw new SimulatorCommandError('No files need merging')
  const configuredTool = String(optionValues(parsed, '--tool').at(-1) ?? state.config?.['merge.tool'] ?? state.operation_metadata?.configured_merge_tool ?? 'tool')
  setOperationMetadata(state, {
    last_mergetool_tool: configuredTool,
    last_mergetool_paths: selected,
    last_mergetool_opened: true,
  })
  return {
    command: 'mergetool',
    stdout: `Opened ${configuredTool} for ${selected.join(', ')}.\nResolve the file in the workspace editor, save it, then run git add for the resolved path.`,
  }
}

export function gitRebase(state: MutableRepositoryState, parsed: ParsedGitCommand): CommandOutcome {
  if (hasOption(parsed, '--abort')) {
    const rebaseState = isRecord(state.rebase_state) ? state.rebase_state : null
    const abortState = rebaseState?.abort_state
    if (!isRecord(abortState)) throw new SimulatorCommandError('fatal: No rebase in progress?')
    for (const key of Object.keys(state)) delete state[key]
    Object.assign(state, clone(abortState))
    setOperationMetadata(state, { last_rebase_aborted: true })
    return { command: 'rebase', stdout: '' }
  }
  if (hasOption(parsed, '--continue')) {
    if (!isRecord(state.rebase_state)) throw new SimulatorCommandError('fatal: No rebase in progress?')
    if (state.conflicts?.length) throw new SimulatorCommandError('You must edit all merge conflicts and then mark them as resolved using git add')
    const remaining = Array.isArray(state.rebase_state.remaining) ? state.rebase_state.remaining.map(String) : []
    if (!remaining.length) {
      setOperationMetadata(state, { last_rebase_new_head: headCommit(state), last_rebase_replayed_commits: clone((state.rebase_state.applied as RepositoryValue[]) ?? []) })
      delete state.rebase_state
      return { command: 'rebase', stdout: 'Successfully rebased and updated current branch.' }
    }
  }
  const ontoValues = optionValues(parsed, '--onto')
  const hasOnto = ontoValues.length > 0
  // Plain `rebase <upstream>`: newbase == upstream == args[0], rebasing the
  // current branch. `rebase --onto <newbase> <upstream> [<branch>]`: replay the
  // <upstream>..<branch> range onto <newbase>; <branch> defaults to HEAD.
  const newBaseRef = String(hasOnto ? ontoValues.at(-1) : parsed.args[0])
  const upstreamRef = String(parsed.args[0])
  const branchRef = hasOnto ? parsed.args[1] : undefined
  const newBase = resolveRef(state, newBaseRef)
  if (!newBase) throw new SimulatorCommandError(`fatal: invalid upstream '${newBaseRef}'`)
  const upstream = resolveRef(state, upstreamRef)
  if (!upstream) throw new SimulatorCommandError(`fatal: invalid upstream '${upstreamRef}'`)
  // With an explicit <branch>, real git checks it out before rebasing so the
  // replayed commits and the moved ref land on that branch, not on prior HEAD.
  if (branchRef) {
    if (!(branchRef in (state.branches ?? {}))) {
      throw new SimulatorCommandError(`fatal: no such branch/commit '${branchRef}'`)
    }
    state.head = { type: 'branch', name: branchRef, target: state.branches?.[branchRef] ?? null }
  }
  const currentBranch = headBranch(state)
  if (!currentBranch) throw new SimulatorCommandError('fatal: rebase requires a branch checkout')
  const branchTip = headCommit(state)
  if (branchTip === newBase) return { command: 'rebase', stdout: 'Current branch is up to date.' }
  // Cut at the merge-base of the branch and the upstream: only commits the branch
  // has beyond the upstream are replayed onto the new base.
  const commitsToReplay = commitsSince(state, branchTip, commonAncestor(state, branchTip, upstream)).reverse()
  let newHead = newBase
  const applied: string[] = []
  for (const sourceId of commitsToReplay) {
    const source = commitById(state, sourceId)
    if (!source) continue
    const base = treeForCommit(state, newHead)
    const nextTree = applyChanges(base, source.changes ?? {})
    const commitId = nextCommitId(state)
    state.commits ??= []
    state.commits.push(commitPayload({ state, commitId, message: source.message ?? `rebase ${sourceId}`, parents: [newHead].filter(Boolean) as string[], tree: nextTree, changes: diffTrees(base, nextTree) }))
    newHead = commitId
    applied.push(commitId)
  }
  setHeadCommit(state, newHead)
  setOperationMetadata(state, {
    last_rebase_target: newBaseRef,
    last_rebase_onto: hasOnto ? newBase : null,
    last_rebase_upstream: upstream,
    last_rebase_new_head: newHead,
    last_rebase_interactive: hasOption(parsed, '-i') || hasOption(parsed, '--interactive'),
    last_rebase_replayed_commits: applied,
  })
  return { command: 'rebase', stdout: `Successfully rebased and updated ${currentBranch}.` }
}
